sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -402,8 +402,9 @@ class CacheManager extends Logging with AdaptiveSparkPlanHelper {
     if (session.conf.get(SQLConf.CAN_CHANGE_CACHED_PLAN_OUTPUT_PARTITIONING)) {
       // Bucketed scan only has one time overhead but can have multi-times benefits in cache,
       // so we always do bucketed scan in a cached plan.
-      SparkSession.getOrCloneSessionWithConfigsOff(
-        session, SQLConf.AUTO_BUCKETED_SCAN_ENABLED :: Nil)
+      SparkSession.getOrCloneSessionWithConfigsOff(session,
+        SQLConf.ADAPTIVE_EXECUTION_APPLY_FINAL_STAGE_SHUFFLE_OPTIMIZATIONS ::
+          SQLConf.AUTO_BUCKETED_SCAN_ENABLED :: Nil)
```

Comment:
I think a question here is: do we want strictly better performance, or generally better performance?

Reply:
Adding an extra shuffle can be very expensive, so I'm inclined to this change.

Reply:
This approach is guaranteed better performance compared to AQE in SQL cache disabled.

Reply:
Thank you, @maryannxue .